The sharded rocks implementation skipped on TSS handling. Backfill.

saintstack opened the pull request:
* fdbserver/MoveKeys.actor.cpp Make startMoveShards work like startMoveKeys handling TSS. Make finishMoveShards work like finishMoveKeys handling TSS.

saintstack commented:
On Testing:

  * Running simulation with SHARD_ENCODE_LOCATION_METADATA=true, and the below change, turned up failures such as   ./build_output/bin/fdbserver -r simulation -f ./build_output/correctness/tests/slow/ddbalance.toml -b off -s 3264324653 . With this PR in place, test this test no longer fails.
  * Ran loadings with TSS enabled along w/ sharded rocks storage engine and tracked metrics across a TSSJointID pairing.

```
diff --git a/fdbserver/SimulatedCluster.actor.cpp b/fdbserver/SimulatedCluster.actor.cpp
index caf0bd9b5..d152a2ca9 100644
--- a/fdbserver/SimulatedCluster.actor.cpp
+++ b/fdbserver/SimulatedCluster.actor.cpp
@@ -2047,9 +2047,7 @@ void SimulationConfig::setProcessesPerMachine(const TestConfig& testConfig) {
 // Also configures the cluster behaviour through setting some flags on the simulator.
 void SimulationConfig::setTss(const TestConfig& testConfig) {
        int tssCount = 0;
-       // TODO: Support TSS in SHARD_ENCODE_LOCATION_METADATA mode.
-       if (!testConfig.simpleConfig && !testConfig.disableTss && !SERVER_KNOBS->SHARD_ENCODE_LOCATION_METADATA &&
-           deterministicRandom()->random01() < 0.25) {
+       if (!testConfig.simpleConfig && !testConfig.disableTss && deterministicRandom()->random01() < 1.0) {
                // 1 or 2 tss
                tssCount = deterministicRandom()->randomInt(1, 3);
        }
```

saintstack commented:
More testing. Added below diff when running 2.5M simulation (suggested by @liquid-helium ):

```
diff --git a/fdbserver/SimulatedCluster.actor.cpp b/fdbserver/SimulatedCluster.actor.cpp
index 784b29e54..96dcb5bda 100644
--- a/fdbserver/SimulatedCluster.actor.cpp
+++ b/fdbserver/SimulatedCluster.actor.cpp
@@ -1644,7 +1644,7 @@ void SimulationConfig::setEncryptionAtRestMode(const TestConfig& testConfig) {
 void SimulationConfig::setStorageEngine(const TestConfig& testConfig) {
        // Using [0, 4) to disable the RocksDB storage engine.
        // TODO: Figure out what is broken with the RocksDB engine in simulation.
-       int storage_engine_type = deterministicRandom()->randomInt(0, 6);
+       int storage_engine_type = 5/*Sharded*/; // deterministicRandom()->randomInt(0, 6);
        if (db.encryptionAtRestMode.isEncryptionEnabled()) {
                // Only storage engine supporting encryption is Redwood.
                storage_engine_type = 3;
```

liquid-helium commented on fdbserver/MoveKeys.actor.cpp:
```diff
@@ -1270,10 +1272,10 @@ ACTOR static Future<Void> startMoveShards(Database occ,
 
-			state Transaction tr(occ);
+			// RYW to optimize re-reading the same key ranges. See readTSSMappingRYW
+			state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(occ);
```
Do we have to use `ReadYourWritesTransaction` here? `Transaction` was used intentionally since `ReadYourWritesTransaction` is not compatible with `unassignServerKeys()`, which is called within `startMoveShards()`, currently it is not actually called since the condition is always `false`, but it could be triggered in the future.

Sorry I should have added a comment here.

saintstack replied:
I was modeling what startMoveKeys does.  The call to readTSSMappingRYW wants a RYWTransaction. Let me go back to plain Transaction (and it looks like readTSSMappingRYW internally doesn't actually need RYWT....).

liquid-helium commented on fdbserver/MoveKeys.actor.cpp:
```diff
@@ -1282,13 +1284,19 @@ ACTOR static Future<Void> startMoveShards(Database occ,
+				wait(checkMoveKeysLock(&(tr->getTransaction()), lock, ddEnabledState));
 
-				Optional<Value> val = wait(tr.get(dataMoveKeyFor(dataMoveId)));
+				if (!loadedTssMapping) {
```
nit: move this `readTSSMappingRYW()` block to line 1335, after we evaluate the `DataMoveMetaData` since we could exit early and end up not using the loaded `TssMapping`.

saintstack replied:
Good one.

liquid-helium commented on fdbserver/MoveKeys.actor.cpp:
```diff
@@ -1746,31 +1759,82 @@ ACTOR static Future<Void> finishMoveShards(Database occ,
 						readyServers.push_back(storageServerInterfaces[s].uniqueID);
 					}
 				}
+				int tssCount = 0;
```
Are we checking if tssCount is the same as the intended tssCount?

saintstack replied:
Again modeling what is done for the move of keys, it does a tssCount just to print. We are just doing same here.

liquid-helium reviewed (commented):
Thanks for implementing this!

saintstack commented:
Address @liquid-helium feedback.

saintstack commented:
Rebased and addressed @liquid-helium  comments.

Reran simulation first with default 100k test count. Two failures. Unrelated.

Next ran 1M simulations w/ below patch in place to favor sharded rocks and tss:

```
diff --git a/fdbclient/ServerKnobs.cpp b/fdbclient/ServerKnobs.cpp
index 6d145c24b..5c889c44d 100644
--- a/fdbclient/ServerKnobs.cpp
+++ b/fdbclient/ServerKnobs.cpp
@@ -168,7 +168,7 @@ void ServerKnobs::initialize(Randomize randomize, ClientKnobs* clientKnobs, IsSi
        init( PRIORITY_ENFORCE_MOVE_OUT_OF_PHYSICAL_SHARD,           960 ); if( randomize && BUGGIFY ) PRIORITY_ENFORCE_MOVE_OUT_OF_PHYSICAL_SHARD = 360; // Set as the lowest priority

        // Data distribution
-       init( SHARD_ENCODE_LOCATION_METADATA,                       false ); if( randomize && BUGGIFY )  SHARD_ENCODE_LOCATION_METADATA = true;
+       init( SHARD_ENCODE_LOCATION_METADATA,                       true ); if( randomize && BUGGIFY )  SHARD_ENCODE_LOCATION_METADATA = true;
        init( ENABLE_DD_PHYSICAL_SHARD,                             false ); // EXPERIMENTAL; If true, SHARD_ENCODE_LOCATION_METADATA must be true; When true, optimization of data move between DCs is disabled
        init( MAX_PHYSICAL_SHARD_BYTES,                         10000000 ); // 10 MB; for ENABLE_DD_PHYSICAL_SHARD; smaller leads to larger number of physicalShard per storage server
        init( PHYSICAL_SHARD_METRICS_DELAY,                        300.0 ); // 300 seconds; for ENABLE_DD_PHYSICAL_SHARD
diff --git a/fdbserver/SimulatedCluster.actor.cpp b/fdbserver/SimulatedCluster.actor.cpp
index caf0bd9b5..d152a2ca9 100644
--- a/fdbserver/SimulatedCluster.actor.cpp
+++ b/fdbserver/SimulatedCluster.actor.cpp
@@ -2047,9 +2047,7 @@ void SimulationConfig::setProcessesPerMachine(const TestConfig& testConfig) {
 // Also configures the cluster behaviour through setting some flags on the simulator.
 void SimulationConfig::setTss(const TestConfig& testConfig) {
        int tssCount = 0;
-       // TODO: Support TSS in SHARD_ENCODE_LOCATION_METADATA mode.
-       if (!testConfig.simpleConfig && !testConfig.disableTss && !SERVER_KNOBS->SHARD_ENCODE_LOCATION_METADATA &&
-           deterministicRandom()->random01() < 0.25) {
+       if (!testConfig.simpleConfig && !testConfig.disableTss && deterministicRandom()->random01() < 1.0) {
                // 1 or 2 tss
                tssCount = deterministicRandom()->randomInt(1, 3);
        }
diff --git a/flow/Knobs.cpp b/flow/Knobs.cpp
index b95ffc9b3..4efc54ba2 100644
--- a/flow/Knobs.cpp
+++ b/flow/Knobs.cpp
@@ -238,7 +238,7 @@ void FlowKnobs::initialize(Randomize randomize, IsSimulated isSimulated) {
        init( ZERO_LENGTH_FILE_PAD,                                  1 );
        init( TRACE_FLUSH_INTERVAL,                               0.25 );
        init( TRACE_RETRY_OPEN_INTERVAL,                                                  1.00 );
-       init( MIN_TRACE_SEVERITY,                isSimulated ?  1 : 10, Atomic::NO ); // Related to the trace severity in Trace.h
+       init( MIN_TRACE_SEVERITY,                isSimulated ?  0 : 10, Atomic::NO ); // Related to the trace severity in Trace.h
        init( MAX_TRACE_SUPPRESSIONS,                              1e4 );
        init( TRACE_DATETIME_ENABLED,                             true ); // trace time in human readable format (always real time)
        init( TRACE_SYNC_ENABLED,                                    0 );
```

30 odd failures... Here is the list sorted and uniq'd:

```
TestFile="tests/fast/AtomicBackupToDBCorrectness.toml"
TestFile="tests/fast/BackupToDBCorrectnessClean.toml"
TestFile="tests/fast/BackupToDBCorrectness.toml"
TestFile="tests/fast/PhysicalShardMove.toml"
TestFile="tests/fast/SpecialKeySpaceRobustness.toml"
TestFile="tests/rare/Throttling.toml"
TestFile="tests/rare/TransactionTagApiCorrectness.toml"
TestFile="tests/restarting/from_7.1.0_until_7.2.0/VersionVectorEnableRestart-1.toml"
TestFile="tests/restarting/from_7.1.0_until_7.2.0/VersionVectorEnableRestart-2.toml"
TestFile="tests/restarting/from_7.3.0/DrUpgradeRestart-1.toml"
TestFile="tests/restarting/from_7.3.0/DrUpgradeRestart-2.toml"
TestFile="tests/restarting/from_7.3.0/UpgradeAndBackupRestore-1.toml"
TestFile="tests/restarting/from_7.3.0/UpgradeAndBackupRestore-2.toml"
TestFile="tests/slow/ApiCorrectnessAtomicRestore.toml"
TestFile="tests/slow/FastTriggeredWatches.toml"
TestFile="tests/slow/ParallelRestoreNewBackupCorrectnessAtomicOp.toml"

```
The TestFile="tests/fast/PhysicalShardMove.toml" looked like it could be related but the complaints were Reason="Consistency check: Shard size is more than 7.000000 std dev from estimate" ... an issue being worked on in another issue.

Identifier for my simulation runs was 20230331-041546-stack-9d22ddb0aa037311

saintstack commented:
Non-change to retriever CI build (failures seem to be related to missing docker deploy found by He Liu)

saintstack commented:
There is a test failure. I just pushed clean main to see if test failure is there too...

saintstack commented:
Four of the above failures are because a run of test_ycsb failed -- the cluster didn't come up:

```
Summarizing 1 Failure:
  [FAIL] [BeforeSuite] 
  /root/go/pkg/mod/github.com/!foundation!d!b/fdb-kubernetes-operator@v1.16.1-0.20230331043336-636ea725bbc4/e2e/fixtures/factory.go:374

Ran 0 of 4 Specs in 4817.370 seconds
FAIL! -- A BeforeSuite node failed so all tests were skipped.
--- FAIL: TestYcsb (4817.37s)
FAIL
FAIL	github.com/FoundationDB/fdb-kubernetes-tests/test_ycsb	4817.415s
FAIL
```

One of the other failures is in simulation because of two fails in  TestFile="tests/restarting/from_7.3.0/DrUpgradeRestart-2.toml"

The other simulation failures are two fails in TestFile="tests/slow/BlobGranuleVerifyLarge.toml"

None of the above looks related. Let me try again running a PR w/ no changes to gauge current state of the build.

liquid-helium commented on fdbserver/MoveKeys.actor.cpp:
```diff
@@ -1446,7 +1453,7 @@ ACTOR static Future<Void> startMoveShards(Database occ,
 
 						// Update dest servers for this range to be equal to servers
-						krmSetPreviouslyEmptyRange(&tr,
+						krmSetPreviouslyEmptyRange(&(tr.getTransaction()),
```
nit: will `&tr` be sufficient?

saintstack replied:
Yes. Was copied from startMoveKeys. Let me fix.

saintstack commented:
The "Result of foundationdb-pr-cluster-tests on Linux CentOS 7" failure is the test_ycsb k8s integration test failing to deploy the cluster in the `BeforeSuite`. I see it on another concurrent PR so likely unrelated: https://github.com/apple/foundationdb/pull/9759

Result of 'Result of foundationdb-pr on Linux CentOS 7' failure is below unit tests failing:

```
	 53 - fdb_c_client_config_tests (Failed)
	 54 - fdb_c_upgrade_from_prev3_gradual (Failed)
	 55 - fdb_c_upgrade_from_prev3_direct (Failed)
	 56 - fdb_c_upgrade_from_prev2_gradual (Failed)
	 57 - fdb_c_upgrade_from_prev2_direct (Failed)
	 58 - fdb_c_upgrade_from_prev (Failed)
	 60 - fdb_c_wiggle_and_upgrade (Failed)
	 61 - fdb_c_shim_library_tests (Failed)
```
